### PR TITLE
Improve support for special characters in blognames

### PR DESCRIPTION
**Pull request by kadamwhite**
I actually think that this patch doesn't fix my problem, but to report the problem: special characters in the blog name are not properly escaped in the sender field of the email dispatched by this plugin. Example,
image

The attached change mirrors how the wordpress/two-factor plugin does things upstream:

https://github.com/WordPress/two-factor/blob/master/providers/class-two-factor-email.php#L227

but I assume it would not impact how the blog name is rendered in what I am guessing is the sender name field. So possibly something that needs to be fixed outside of this plugin.

### CODE REVIEW

**Comment by ajvillegas** [comment]
@kadamwhite I took the liberty of updating this PR in order to fix humanmade/playstation/issues/3130.

Using `wp_specialchars_decode()` does not solve the issue on AWS. Quoting @rmccue on issue humanmade/aws-ses-wp-mail/issues/46:

[According to] the [AWS SendEmail endpoint specifically notes](https://docs.aws.amazon.com/ses/latest/APIReference/API_SendEmail.html):

> The sender name (also known as the *friendly name*) may contain non-ASCII characters. These characters must be encoded using MIME encoded-word syntax, as described in RFC 2047. MIME encoded-word syntax uses the following form: `=?charset?encoding?encoded-text?=`.

I tested the functionality on Local Server and it does works as expected, also fixing issue humanmade/playstation/issues/3130. Here is a screenshot of the email subjects for Spanish and Traditional Chinese on Mailhog:

Screen Shot 2023-01-12 at 7 34 59 PM

File: providers/class.two-factor-email.php [hunk 1]
```diff
@@ -148,7 +148,7 @@ public function generate_and_email_token( $user ) {
 		$token = $this->generate_token( $user->ID );
 
 		/* translators: %s: site name */
-		$subject = wp_strip_all_tags( sprintf( __( 'Your login confirmation code for %s', 'two-factor' ), get_bloginfo( 'name' ) ) );
+		$subject = wp_strip_all_tags( sprintf( __( 'Your login confirmation code for %s', 'two-factor' ), mb_encode_mimeheader( get_option( 'blogname' ), ENT_QUOTES ) ) );
```
Comment (roborourke) [inline providers/class.two-factor-email.php:151]: What about `html_entity_decode()`? Looks fine if it works as expected though.

**Review by roborourke — APPROVED** [review]

File: providers/class.two-factor-email.php (hunk 1 above)
Comment (kadamwhite) [reply providers/class.two-factor-email.php:151]: I'd defer to @ajvillegas on whether that was tried or not. `wp_specialchars_decode` which the upstream uses definitely does _not_ fix it, so I would guess that the more specific function here is likely necessary.

**Comment by kadamwhite** [comment]
@roborourke This had worked locally but when I installed on a dev stack just now, I got this fatal when trying to initiate an email code resend:

```
Uncaught ValueError: mb_encode_mimeheader(): Argument #2 ($charset) must be a valid encoding, "3" given in /usr/src/app/vendor/humanmade/two-factor/providers/class.two-factor-email.php:151
Stack trace:
#0 /usr/src/app/vendor/humanmade/two-factor/providers/class.two-factor-email.php(151): mb_encode_mimeheader('Blo...', '3')
#1 /usr/src/app/vendor/humanmade/two-factor/providers/class.two-factor-email.php(171): Two_Factor_Email->generate_and_email_token(Object(WP_User))
#2 /usr/src/app/vendor/humanmade/two-factor/class.two-factor-core.php(361): Two_Factor_Email->authentication_page(Object(WP_User))
#3 /usr/src/app/vendor/humanmade/two-factor/class.two-factor-core.php(303): Two_Factor_Core::login_html(Object(WP_User), '0d498aca32a1c7d...', 'https://site...', '', Object(Two_Factor_Email))
#4 /usr/src/app/wordpress/wp-includes/class-wp-hook.php(307): Two_Factor_Core::backup_2fa('')
#5 /usr/src/app/wordpress/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
#6 /usr/src/app/wordpress/wp-includes/plugin.php(474): WP_Hook->do_action(Array)
#7 /usr/src/app/wordpress/wp-login.php(518): do_action('login_form_back...')
#8 {main}
  thrown
```
I recommend we untag and un-publish 0.3.1 for now

**Comment by ajvillegas** [comment]
@kadamwhite The `mb_encode_mimeheader` function is [expecting the `charset` as the second parameter](https://www.php.net/manual/en/function.mb-encode-mimeheader.php#refsect1-function.mb-encode-mimeheader-parameters), but I left the `ENT_QUOTES` flag from the `wp_specialchars_decode` we had previously on the PR and think this is what's causing the error. It looks like the second parameter should be `UFT-8`.

**Comment by kadamwhite** [comment]
Reverted in 59b577a and will open a new PR to iterate

**Pull request merged**